// web-bluetooth-mock/dist/index.js
"use strict";
// Web Bluetooth Mock
//
// Copyright (C) 2017, Uri Shaked
// Published under the terms of the MIT license
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
/// <reference path="./event-target.d.ts" />
var event_target_shim_1 = require("event-target-shim");
function leftPad(s, count, pad) {
    while (s.length < count) {
        s = pad + s;
    }
    return s;
}
function normalizeUuid(service) {
    if (typeof service === "number" && service > 0) {
        return leftPad(service.toString(16), 8, "0") + "-0000-1000-8000-00805f9b34fb";
    }
    else {
        // TODO: handle standard UUID names, throw exception on invalid values
        return service.toString();
    }
}
var CharacteristicMock = /** @class */ (function (_super) {
    __extends(CharacteristicMock, _super);
    function CharacteristicMock(service, uuid) {
        var _this = _super.call(this) || this;
        _this.service = service;
        _this.uuid = uuid;
        _this.properties = {
            authenticatedSignedWrites: false,
            broadcast: false,
            indicate: false,
            notify: false,
            read: false,
            reliableWrite: false,
            writableAuxiliaries: false,
            write: false,
            writeWithoutResponse: false,
        };
        _this.value = new DataView(new Uint8Array(0).buffer);
        return _this;
    }
    CharacteristicMock.prototype.startNotifications = function () {
        return Promise.resolve(this);
    };
    CharacteristicMock.prototype.stopNotifications = function () {
        return Promise.resolve(this);
    };
    CharacteristicMock.prototype.readValue = function () {
        return Promise.resolve(this.value);
    };
    CharacteristicMock.prototype.writeValue = function () {
        return Promise.resolve();
    };
    CharacteristicMock.prototype.getDescriptor = function () {
        return Promise.reject(new Error("Not implemented"));
    };
    CharacteristicMock.prototype.getDescriptors = function () {
        return Promise.reject(new Error("Not implemented"));
    };
    return CharacteristicMock;
}(event_target_shim_1.EventTarget));
exports.CharacteristicMock = CharacteristicMock;
var PrimaryServiceMock = /** @class */ (function (_super) {
    __extends(PrimaryServiceMock, _super);
    function PrimaryServiceMock(device, uuid, isPrimary) {
        if (isPrimary === void 0) { isPrimary = true; }
        var _this = _super.call(this) || this;
        _this.device = device;
        _this.uuid = uuid;
        _this.isPrimary = isPrimary;
        _this.characteristicMocks = {};
        return _this;
    }
    PrimaryServiceMock.prototype.getCharacteristic = function (characteristic) {
        return Promise.resolve(this.getCharacteristicMock(characteristic));
    };
    PrimaryServiceMock.prototype.getCharacteristics = function () {
        var _this = this;
        return Promise.resolve(Object.keys(this.characteristicMocks)
            .map(function (k) { return _this.characteristicMocks[k]; }));
    };
    PrimaryServiceMock.prototype.getIncludedService = function () {
        return Promise.reject(new Error("Not implemented"));
    };
    PrimaryServiceMock.prototype.getIncludedServices = function () {
        return Promise.resolve([]);
    };
    PrimaryServiceMock.prototype.getCharacteristicMock = function (characteristic) {
        if (!this.characteristicMocks[characteristic]) {
            this.characteristicMocks[characteristic] = new CharacteristicMock(this, normalizeUuid(characteristic));
        }
        return this.characteristicMocks[characteristic];
    };
    return PrimaryServiceMock;
}(event_target_shim_1.EventTarget));
exports.PrimaryServiceMock = PrimaryServiceMock;
var GattMock = /** @class */ (function (_super) {
    __extends(GattMock, _super);
    function GattMock(device) {
        var _this = _super.call(this) || this;
        _this.device = device;
        _this.connected = false;
        _this.device = device;
        return _this;
    }
    GattMock.prototype.connect = function () {
        this.connected = true;
        return Promise.resolve(this);
    };
    GattMock.prototype.disconnect = function () {
        this.connected = false;
    };
    GattMock.prototype.getPrimaryService = function (service) {
        return Promise.resolve(this.device.getServiceMock(service));
    };
    GattMock.prototype.getPrimaryServices = function (service) {
        if (service) {
            this.device.getServiceMock(service);
        }
        return Promise.resolve(this.device.getServiceMocks());
    };
    return GattMock;
}(event_target_shim_1.EventTarget));
exports.GattMock = GattMock;
var DeviceMock = /** @class */ (function (_super) {
    __extends(DeviceMock, _super);
    function DeviceMock(name, services) {
        var _this = _super.call(this) || this;
        _this.name = name;
        _this.services = services;
        _this.watchingAdvertisements = false;
        _this.id = "";
        _this.serviceMocks = {};
        _this.gatt = new GattMock(_this);
        return _this;
    }
    DeviceMock.prototype.hasService = function (service) {
        return this.services && this.services.indexOf(service) >= 0;
    };
    DeviceMock.prototype.getServiceMock = function (service) {
        if (!this.serviceMocks[service]) {
            this.serviceMocks[service] = new PrimaryServiceMock(this, normalizeUuid(service));
        }
        return this.serviceMocks[service];
    };
    DeviceMock.prototype.getServiceMocks = function () {
        var _this = this;
        return Object.keys(this.serviceMocks).map(function (k) { return _this.serviceMocks[k]; });
    };
    DeviceMock.prototype.watchAdvertisements = function () {
        this.watchingAdvertisements = true;
        return Promise.resolve();
    };
    DeviceMock.prototype.unwatchAdvertisements = function () {
        this.watchingAdvertisements = false;
    };
    return DeviceMock;
}(event_target_shim_1.EventTarget));
exports.DeviceMock = DeviceMock;
var WebBluetoothMock = /** @class */ (function () {
    function WebBluetoothMock(devices) {
        this.devices = devices;
    }
    WebBluetoothMock.prototype.requestDevice = function (options) {
        for (var _i = 0, _a = this.devices; _i < _a.length; _i++) {
            var device = _a[_i];
            for (var _b = 0, _c = options.filters; _b < _c.length; _b++) {
                var filter = _c[_b];
                if (filter.name && filter.name === device.name) {
                    return Promise.resolve(device);
                }
                if (filter.namePrefix && device.name && device.name.indexOf(filter.namePrefix) === 0) {
                    return Promise.resolve(device);
                }
                if (filter.services) {
                    var found = true;
                    for (var _d = 0, _e = filter.services; _d < _e.length; _d++) {
                        var service = _e[_d];
                        found = found && device.hasService(service);
                    }
                    if (found) {
                        return Promise.resolve(device);
                    }
                }
            }
        }
        return Promise.reject(new Error("User cancelled device chooser"));
    };
    return WebBluetoothMock;
}());
exports.WebBluetoothMock = WebBluetoothMock;
