// fuse-box/dist/commonjs/quantum/plugin/ResponsiveAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Utils_1 = require("../../Utils");
const Config_1 = require("../../Config");
class ResponsiveAPI {
    constructor(core) {
        this.core = core;
        this.computedStatements = false;
        this.hashes = false;
        this.isServerFunction = false;
        this.isBrowserFunction = false;
        this.customMappings = {};
        this.customStatementResolve = false;
    }
    addComputedRequireStatetements() {
        this.computedStatements = true;
        this.hashes = true;
    }
    hashesUsed() {
        return this.hashes;
    }
    addMapping(fuseBoxPath, id) {
        this.customMappings[fuseBoxPath] = id;
        this.customStatementResolve = true;
    }
    addIsServerFunction() {
        this.isServerFunction = true;
    }
    addIsBrowserFunction() {
        this.isBrowserFunction = true;
    }
    render() {
        const options = {
            browser: this.core.opts.isTargetBrowser(),
            universal: this.core.opts.isTargetUniveral(),
            server: this.core.opts.isTargetServer(),
            isServerFunction: this.isServerFunction,
            isBrowserFunction: this.isBrowserFunction,
            computedStatements: this.computedStatements,
            hashes: this.hashes,
            customStatementResolve: this.customStatementResolve
        };
        const variables = {};
        if (Object.keys(this.customMappings).length > 0) {
            variables.customMappings = this.customMappings;
        }
        return Utils_1.jsCommentTemplate(path.join(Config_1.Config.FUSEBOX_MODULES, "fuse-box-responsive-api/index.js"), options, variables);
    }
}
exports.ResponsiveAPI = ResponsiveAPI;
